src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp

```diff
@@ -197,63 +196,73 @@ bool AsynchronousReadIndirectBufferFromRemoteFS::nextImpl()
 }
 
 
-off_t AsynchronousReadIndirectBufferFromRemoteFS::seek(off_t offset_, int whence)
+off_t AsynchronousReadIndirectBufferFromRemoteFS::seek(off_t offset, int whence)
 {
     ProfileEvents::increment(ProfileEvents::RemoteFSSeeks);
 
-    if (whence == SEEK_CUR)
+    size_t new_pos;
+    if (whence == SEEK_SET)
     {
-        /// If position within current working buffer - shift pos.
-        if (!working_buffer.empty() && static_cast<size_t>(getPosition() + offset_) < file_offset_of_buffer_end)
-        {
-            pos += offset_;
-            return getPosition();
-        }
-        else
-        {
-            file_offset_of_buffer_end += offset_;
-        }
+        assert(offset >= 0);
+        new_pos = offset;
+    }
+    else if (whence == SEEK_CUR)
+    {
+        new_pos = file_offset_of_buffer_end - (working_buffer.end() - pos) + offset;
     }
-    else if (whence == SEEK_SET)
+    else
+    {
+        throw Exception("ReadBufferFromFileDescriptor::seek expects SEEK_SET or SEEK_CUR as whence", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+    }
+
+    /// Position is unchanged.
+    if (new_pos + (working_buffer.end() - pos) == file_offset_of_buffer_end)
+        return new_pos;
+
+    bool read_from_prefetch = false;
+    while (true)
     {
-        /// If position is within current working buffer - shift pos.
-        if (!working_buffer.empty()
-            && static_cast<size_t>(offset_) >= file_offset_of_buffer_end - working_buffer.size()
-            && size_t(offset_) < file_offset_of_buffer_end)
+        if (file_offset_of_buffer_end - working_buffer.size() <= static_cast<size_t>(new_pos)
+            && new_pos <= file_offset_of_buffer_end)
         {
-            pos = working_buffer.end() - (file_offset_of_buffer_end - offset_);
+            /// Position is still inside the buffer.
+            /// Probably it is at the end of the buffer - then we will load data on the following 'next' call.
 
+            pos = working_buffer.end() - file_offset_of_buffer_end + new_pos;
             assert(pos >= working_buffer.begin());
             assert(pos <= working_buffer.end());
 
-            return getPosition();
+            return new_pos;
         }
-        else
+        else if (prefetch_future.valid())
         {
-            file_offset_of_buffer_end = offset_;
+            /// Read from prefetch buffer and recheck if the new position is valid inside.
+
+            if (nextImpl())
+            {
+                read_from_prefetch = true;
+                continue;
+            }
```

Comment:
But it means that if seek is not within current working buffer, then we always wait for prefetch, and I think this can lead to slow down because seek is also done when calling `MergeTreeReaderWide::prefetch()`.